File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/MediaFolders.java
package in.org.quicko.silo.client.media;

import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.RequestOptions;
import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.JsonValues;
import in.org.quicko.silo.client.transport.Transport;
import in.org.quicko.silo.client.transport.TransportRequest;
import java.util.List;
import java.util.Map;

/**
 * The library's folder tree: {@code list}, {@code create}, {@code rename} and
 * {@code delete}. A folder is an explicit record, so one can exist before
 * anything is filed into it.
 */
public final class MediaFolders {
  private final Transport transport;

  public MediaFolders(Transport transport) {
    this.transport = transport;
  }

  public List<String> list() {
    return list(RequestOptions.none());
  }

  public List<String> list(RequestOptions options) {
    JsonNode body = transport.json(
        TransportRequest.get(ApiPath.mediaFolders()).options(options).build());
    return JsonValues.strings(body.get("items"));
  }

  public String create(String path) {
    return create(path, RequestOptions.none());
  }

  /** Answers the path the server settled on. */
  public String create(String path, RequestOptions options) {
    JsonNode body = transport.json(
        TransportRequest.post(ApiPath.mediaFolders())
            .body(Map.of("path", path))
            .options(options)
            .build());
    return JsonValues.text(body, "path");
  }

  public MediaFolderRename rename(String from, String to) {
    return rename(from, to, MediaFolderMoveOptions.none());
  }

  /** Renames or moves a folder, and every asset in the subtree with it. No
   *  entry is touched and no blob moves: an asset is addressed by its id. */
  public MediaFolderRename rename(String from, String to, MediaFolderMoveOptions options) {
    return MediaFolderRename.fromWire(transport.json(
        TransportRequest.patch(ApiPath.mediaFolders())
            .body(Map.of("from", from, "to", to, "merge", options.merge()))
            .options(options.request())
            .build()));
  }

  public void delete(String path) {
    delete(path, MediaFolderDeleteOptions.none());
  }

  /** The folder path travels as {@code ?path=} rather than a path segment,
   *  because it contains slashes of its own. */
  public void delete(String path, MediaFolderDeleteOptions options) {
    transport.empty(
        TransportRequest.delete(ApiPath.mediaFolders())
            .query("path", path)
            .query("recursive", options.recursive() ? true : null)
            .query("force", options.force() ? true : null)
            .options(options.request())
            .build());
  }
}

File: packages/silo-client-dart/lib/src/media/media_folders.dart
import '../request_options.dart';
import '../transport/api_path.dart';
import '../transport/json_values.dart';
import '../transport/transport.dart';
import '../transport/transport_request.dart';
import 'media_folder_delete_options.dart';
import 'media_folder_move_options.dart';

/// The folder tree. A folder is a record, so it can exist while empty.
final class MediaFolders {
  const MediaFolders(this._transport);

  final Transport _transport;

  Future<List<String>> list([RequestOptions options = const RequestOptions()]) async => JsonValues.strings(
    (await _transport.json(TransportRequest('GET', ApiPath.mediaFolders(), options: options)))['items'],
  );

  /// Answers the path the server settled on.
  Future<String> create(String path, [RequestOptions options = const RequestOptions()]) async =>
      (await _transport.json(
            TransportRequest('POST', ApiPath.mediaFolders(), body: {'path': path}, options: options),
          ))['path']
          as String;

  Future<({String from, String to})> rename(
    String from,
    String to, [
    MediaFolderMoveOptions options = const MediaFolderMoveOptions(),
  ]) async {
    final body = await _transport.json(
      TransportRequest(
        'PATCH',
        ApiPath.mediaFolders(),
        body: {'from': from, 'to': to, 'merge': options.merge},
        options: options,
      ),
    );
    return (from: body['from'] as String, to: body['to'] as String);
  }

  /// The path travels as `?path=`, since it contains `/`.
  Future<void> delete(String path, [MediaFolderDeleteOptions options = const MediaFolderDeleteOptions()]) =>
      _transport.empty(
        TransportRequest(
          'DELETE',
          ApiPath.mediaFolders(),
          query: {'path': path, if (options.recursive) 'recursive': true, if (options.force) 'force': true},
          options: options,
        ),
      );
}
